### PR TITLE
KH2 game implementation

### PR DESCRIPTION
**Pull request by JaredWeakStrike**
Please format your title with what portion of the project this pull request is
targeting and what it's changing.

ex. "MyGame4: implement new game" or "Docs: add new guide for customizing MyGame3"

## What is this fixing or adding? 
Adding base support for Kingdom Hearts 2 in Archipelago

## How was this tested? 
over 30 seeds with every setting with most games that are supported

### CODE REVIEW

File: worlds/kh2/__init__.py [hunk 1]
```diff
@@ -0,0 +1,310 @@
...
+            "setup_en.md",
+            "setup/en",
+            ["JaredWeakStrike"]
+    )]
+
+
+class KH2World(World):
+    game: str = "Kingdom Hearts 2"
+
+    data_version = 0
```
Comment (kindasneaki) [inline worlds/kh2/__init__.py:30]: This should be 1 for the release.

**Comment by kindasneaki** [comment]
You are missing setup docs, and a game page doc for the webhost.

File: worlds/sa2b/__init__.py [hunk 2]
```diff
@@ -318,7 +318,7 @@ def create_regions(self):
 
     def create_item(self, name: str, force_non_progression=False) -> Item:
         data = item_table[name]
-
+
```
[the remark this answers is not in the document]
Comment (JaredWeakStrike) [reply worlds/sa2b/__init__.py]: resolved

File: WebHostLib/upload.py [hunk 3]
```diff
@@ -68,7 +68,6 @@ def upload_zip_to_db(zfile: zipfile.ZipFile, owner=None, meta={"race": False}, s
             _, _, slot_id, *_ = file.filename.split('_')[0].split('-', 3)
             data = zfile.open(file, "r").read()
             files[int(slot_id[1:])] = data
-
```
[the remark this answers is not in the document]
Comment (JaredWeakStrike) [reply WebHostLib/upload.py]: resolved

File: worlds/kh2/logic.py [hunk 4]
```diff
@@ -0,0 +1,107 @@
+from .Names import ItemName
+from ..AutoWorld import LogicMixin
+
+
+class KH2Logic(LogicMixin):
+    # piglets house
+    def kh_torn_page_1(self, player: int):
```
[the remark this answers is not in the document]
Comment (JaredWeakStrike) [reply worlds/kh2/logic.py]: yeah I totally forgot I had that in here lol. Resolved now

File: worlds/kh2/__init__.py (hunk 1 above)
Comment (JaredWeakStrike) [reply worlds/kh2/__init__.py]: resolved

File: KH2Client.py [hunk 5]
```diff
@@ -0,0 +1,554 @@
+import os
+import asyncio
+import ModuleUpdate
+import typing
+import json
+import Utils
+from pymem import pymem
```
Comment (Berserker66) [inline KH2Client.py:7]: I don't see this dependency listed.

Comment (JaredWeakStrike) [reply KH2Client.py:7]: It is now apart of requirements.txt

File: requirements.txt [hunk 6]
```diff
@@ -5,4 +5,5 @@ jellyfish>=0.9.0
 jinja2>=3.1.2
 schema>=0.7.5
 kivy>=2.1.0
-bsdiff4>=1.2.2
\ No newline at end of file
+bsdiff4>=1.2.2
+Pymem>=1.10.0
```
Comment (Berserker66) [inline requirements.txt:9]: your world specific requirements go into your world folder's requirements.txt

File: worlds/kh2/Options.py [hunk 7]
```diff
@@ -0,0 +1,241 @@
+from Options import Choice, Option, Range, Toggle, OptionSet
+import typing
+
+
+class SoraEXP(Range):
+    """Sora Level Exp Multiplier"""
+    display = "Sora Level EXP"
```
Comment (Berserker66) [inline worlds/kh2/Options.py:7]:
```suggestion
    display_name = "Sora Level EXP"
```

File: worlds/kh2/Options.py [hunk 8]
```diff
@@ -0,0 +1,241 @@
...
+    display_name = "Random Visit Locking Item"
+    range_start=0
+    range_end=27
+    default=3
+
+class SuperBosses(Toggle):
+    """Terra, Sephiroth and Data Fights Toggle."""
+    display_name = "Super Bosses"
+    default = False
+class Cups(Choice):
```
Comment (Berserker66) [inline worlds/kh2/Options.py:117]:
```suggestion
    default = False

class Cups(Choice):
```

File: worlds/kh2/__init__.py (hunk 1 above)
Comment (Berserker66) [reply worlds/kh2/__init__.py:36]: is it?

File: worlds/kh2/__init__.py [hunk 9]
```diff
@@ -0,0 +1,314 @@
...
+        while len(self.goofy_ability_pool) < 33:
+            self.goofy_ability_pool.append(self.multiworld.per_slot_randoms[self.player].choice(self.goofy_ability_pool))
+
+    def generate_basic(self):
+        itempool: typing.List[KH2Item] = []
+
+        self.hitlist = list()
+        self.filler_items.extend(item_groups["Filler"])
+
+        if self.multiworld.FinalXemnas[self.player].value == 1:
```
Comment (Berserker66) [inline worlds/kh2/__init__.py:191]:
```suggestion
        if self.multiworld.FinalXemnas[self.player] == 1:
```
all the options support direct comparison, with either the option key, option name str or option int and will check that those options are even plausible

**Comment by JaredWeakStrike** [comment]
fixed everything from the review and fixed item duping.

File: requirements.txt (hunk 6 above)
Comment (JaredWeakStrike) [reply requirements.txt]: resolved

File: worlds/kh2/Options.py (hunk 7 above)
Comment (JaredWeakStrike) [reply worlds/kh2/Options.py]: resolved

File: worlds/kh2/Options.py (hunk 8 above)
Comment (JaredWeakStrike) [reply worlds/kh2/Options.py]: resolved

File: worlds/kh2/__init__.py (hunk 9 above)
Comment (JaredWeakStrike) [reply worlds/kh2/__init__.py]: resolved

File: worlds/kh2/__init__.py [hunk 10]
```diff
@@ -0,0 +1,314 @@
...
+            self.visitlockingitem.remove(item)
+
+        # there are levels but level 1 is there to keep code clean
+        if self.multiworld.LevelDepth[self.player].value == 3:
+            # level 99 sanity
+            self.totalLocations -= 1
+        elif self.multiworld.LevelDepth[self.player].value == 2:
+            # level 50 sanity
+            self.totalLocations -= 50
+        elif self.multiworld.LevelDepth[self.player].value == 4:
```
[the remark this answers is not in the document]
Comment (JaredWeakStrike) [reply worlds/kh2/__init__.py]: resolved

**Comment by JaredWeakStrike** [comment]
okay it is now actually ready for merge

**Pull request merged**